### ISSUE TITLE
False report on InstrumentedAttribute (sqlalchemy field) is not a hashable

### ISSUE THREAD

**Issue by ManiMozaffar** [microsoft/pyright#7462]
**I'm gonna make a PR to sqlalchemy, and fix this. The intention of this issue is more of introducing a way to spefiecally ignore a very certain type error.**

InstrumentedAttribute is hashable. I think i understand why vscode can't understand that this is not hashable, but what I don't understand is why this feature was introduced in first place. Python is way too magical. That would be nice if we can have a way to disable this in type check. Only this error, not other types of reportGeneralTypeIssues
Right now i need to add pyright: ignore [reportGeneralTypeIssues] to many places in code, which is not very ideal. Or even worse, ignore that on config. So let's get into issue and reproduce it; Install sqlalchemy:

```python
from typing import Any

from sqlalchemy.orm import (
    DeclarativeBase,
    InstrumentedAttribute,
    Mapped,
    MappedAsDataclass,
    mapped_column,
)

class BaseSQL(MappedAsDataclass, DeclarativeBase):
    ...


class Model(BaseSQL):
    __tablename__ = "Model"
    id: Mapped[int] = mapped_column(primary_key=True)


result: dict[InstrumentedAttribute, Any] = {Model.id: ...}
print(Model.id.__hash__())
assert Model.id.__hash__() is not None # gonna run just fine :) 
```

Can we ignore it? sure... **but at least that would be nice to be able to specifically ignore this type** there's no straightforward way of saying if an object is hashable or not. You can't just say because it has  `__eq__` then it's no longer hashable. what if i write the hash function myself and monkey patch it with metaclass? We can also achieve this by using Hashable from python typing, if we need to, right? `dict[Hashable, ...]`
When monkey patch I mean this:

```python
class AutoHashMeta(type):
    def __new__(mcs, name, bases, namespace, **kwds):
        cls = super().__new__(mcs, name, bases, dict(namespace))

        if "__eq__" in namespace and "__hash__" not in namespace:

            def __hash__(self):
                return hash((type(self),) + tuple(self.__dict__.values()))

            cls.__hash__ = __hash__

        return cls


class Test(metaclass=AutoHashMeta):
    def __init__(self, value: Any):
        self.value = value

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, type(self)):
            return NotImplemented
        return self.value == other.value

```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?

pyright 1.1.335
❯ code --version
1.87.1

**Comment by erictraut** [comment]
I'm not able to repro any hashing-related type errors with the code above. Perhaps it's specific to a particular version of sqlalchemy? I've installed version 2.0.28, which I think is the latest published version. Do you have any type stubs installed for sqlalchemy?

**Comment by erictraut** [comment]
In any case, I agree that the hashability check should be part of its own diagnostic rule rather than lumped in with reportGeneralTypeIssue. I've added a new rule called `reportUnhashable` that will control this diagnostic in the next version.

**Comment by ManiMozaffar** [comment]
<img width="855" alt="Screenshot 2024-03-12 at 07 55 18" src="https://github.com/microsoft/pyright/assets/86913668/bb1f5f5e-9691-4ec3-afaa-4aaf1e3eeb1c">
Name: SQLAlchemy
Version: 2.0.25
I don't have any type stubs installed for sqlalchemy. It also happen in 2 of my repos, but not in one, which is weird. I spent hours debugging this, but couldn't figure it out why. but now that you said it...

<img width="669" alt="Screenshot 2024-03-12 at 07 58 32" src="https://github.com/microsoft/pyright/assets/86913668/d75cdd2e-dc1d-45a1-ab9c-9927da44026a">
I ran the pyright command and it's not triggered by Vscode. does it have to do with pylance?

**Comment by ManiMozaffar** [comment]
My bad, this was introduced in 1.1.334 or 1.1.335
<img width="880" alt="Screenshot 2024-03-12 at 08 08 50" src="https://github.com/microsoft/pyright/assets/86913668/0933cd7d-0697-4f4a-888d-58b7151c63e1">

**Comment by ManiMozaffar** [comment]
how do I have a version installed that is not released on pypi? 

https://pypi.org/project/pyright/#history

<img width="564" alt="Screenshot 2024-03-12 at 08 16 01" src="https://github.com/microsoft/pyright/assets/86913668/5efa71dd-3df3-48b3-9fbe-2c7f77f561d4">

**Comment by erictraut** [comment]
Pyright 1.1.335 is quite old. Please upgrade to 1.1.353, which is the latest.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.354, which I just published.
